src/front/glsl/ast.rs

```diff
@@ -141,99 +142,112 @@ pub enum Profile {
 
 #[derive(Debug)]
 pub struct Context<'function> {
-    pub expressions: &'function mut Arena<Expression>,
+    expressions: &'function mut Arena<Expression>,
     pub locals: &'function mut Arena<LocalVariable>,
     pub arguments: &'function mut Vec<FunctionArgument>,
 
     //TODO: Find less allocation heavy representation
     pub scopes: Vec<FastHashMap<String, VariableReference>>,
     pub lookup_global_var_exps: FastHashMap<String, VariableReference>,
-    pub lookup_constant_exps: FastHashMap<String, VariableReference>,
-    pub typifier: Typifier,
     pub samplers: FastHashMap<Handle<Expression>, Handle<Expression>>,
+    pub typifier: Typifier,
 
     pub hir_exprs: Arena<HirExpr>,
+    emitter: Emitter,
 }
 
 impl<'function> Context<'function> {
     pub fn new(
+        program: &mut Program,
+        body: &mut Block,
         expressions: &'function mut Arena<Expression>,
         locals: &'function mut Arena<LocalVariable>,
         arguments: &'function mut Vec<FunctionArgument>,
     ) -> Self {
-        Context {
+        let mut this = Context {
             expressions,
             locals,
             arguments,
 
             scopes: vec![FastHashMap::default()],
-            lookup_global_var_exps: FastHashMap::default(),
-            lookup_constant_exps: FastHashMap::default(),
+            lookup_global_var_exps: FastHashMap::with_capacity_and_hasher(
+                program.constants.len() + program.global_variables.len(),
+                Default::default(),
+            ),
             typifier: Typifier::new(),
             samplers: FastHashMap::default(),
 
             hir_exprs: Arena::default(),
-        }
-    }
+            emitter: Emitter::default(),
+        };
 
-    pub fn lookup_local_var(&self, name: &str) -> Option<VariableReference> {
-        for scope in self.scopes.iter().rev() {
-            if let Some(var) = scope.get(name) {
-                return Some(var.clone());
-            }
-        }
-        None
-    }
+        this.emit_start();
 
-    pub fn lookup_global_var(
-        &mut self,
-        program: &mut Program,
-        name: &str,
-    ) -> Option<VariableReference> {
-        self.lookup_global_var_exps.get(name).cloned().or_else(|| {
-            let expr = match *program.lookup_global_variables.get(name)? {
+        for &(ref name, lookup) in program.global_variables.iter() {
+            let expr = match lookup {
                 GlobalLookup::Variable(v) => Expression::GlobalVariable(v),
                 GlobalLookup::BlockSelect(handle, index) => {
-                    let base = self.expressions.append(Expression::GlobalVariable(handle));
+                    let base = this.add_expression(Expression::GlobalVariable(handle), body);
```

Comment:
Here we know that everything in this loop will be implicitly emitted (the global variables and accessing into them, so maybe we can just do `this.expressions.push()` instead of relying on the emitter?

Reply:
Access are implicitly emitted?

Reply:
accesses that result in pointers don't need to be emitted, since there is always going to be either a Store into them, or a Load (which is emitted)